zipextimporter: improves compatibility

SeaHOH opened the pull request:
Add support import .pyd packages, e.g. `name\\__init__.pyd.`, also makes modules search order as same as non-zip.

Add support import pywin32 DLLs.

Add support compatible with export hook name (PEP 489).
https://peps.python.org/pep-0489/#export-hook-name

Add some missing methods.

Tweak install function.

Minimum compatible to CPython 36.

SeaHOH commented:
Export hook name is untested.

SeaHOH commented:
Fix import/search order, make it as same as non-zip:
- `package` > `non-package`
- `.pyd` > `.pyc` > `.py`

SeaHOH commented:
Add missing doc.

albertosottile commented:
Thanks for this contribution. I need to review this carefully, as the change is not trivial, but I am quite swamped at the moment and so this will take a while.

albertosottile commented on zipextimporter.py:
```diff
@@ -42,138 +43,227 @@
+        return super().__repr__().replace("zipimporter", "ZipExtensionImporter")
+
+
+def install(hook=hasattr(zipimporter, "_files")):
```
I am a little confused... the class `zipimporter` has never an attribute `_files`. So, what is the difference between this and `hook=False`? 

In addition, could you comment a little on the two installation methods? Why was the old one (`_install_hook`) not good enough? Why monkey patching `zipimport` should be better? I find inserting in the `path_hooks` is much more elegant and stable than messing with code that we cannot truly control...

SeaHOH replied:
There is no difference, `hasattr(...)` expression was kept as memento. If unadvisable, please remove it.

I think monkey patch `zipimport.zipimporter` is better way, will not break codes which base on  `zipimport.zipimporter`. Because `zipextimporter` is not a official lib module, there is no compatibility at all.

albertosottile replied:
I do not understand this. In the current implementation, code that is based on `zipimporter` will work out-of-the-box provided that the author changes the way they initialize the importer. There should be no need for monkey patching the original code.

albertosottile commented on zipextimporter.py:
```diff
@@ -42,138 +43,227 @@
+
+
+class ZipExtensionImporter(zipimporter):
+    suffixes = _imp.extension_suffixes()
+    _searchorder = (
+        *[(f"\\__init__{suffix}", True, True) for suffix in suffixes],
+        ("\\__init__.pyc", False, True),
+        ("\\__init__.py", False, True),
+        *[(suffix, True, False) for suffix in suffixes],
+        (".pyc", False, False),
+        (".py", False, False),
+    )
+    _searchorder_pywin32 = (
+        *_searchorder[:-2],
+        ("%d%d.dll" % sys.version_info[:2], True, False),
+        *_searchorder[-2:],
+    )
+    del suffixes
+    names_pywin32 = frozenset(["pywintypes", "pythoncom"])
+    verbose = _memimporter.get_verbose_flag()
+
```
All this is a little too much logic to happen at class declaration time, I would move all this in `__init__`.

SeaHOH replied:
I put them here because `_suffixes` is at here before. I suggest move them to `global` except `verbose`.

albertosottile commented on zipextimporter.py:
```diff
@@ -42,138 +43,227 @@
+        ("%d%d.dll" % sys.version_info[:2], True, False),
+        *_searchorder[-2:],
+    )
+    del suffixes
```
I am not a fan of `del` so I would use it only if it is absolutely necessary.

SeaHOH replied:
Agree with you, this is not necessary.

albertosottile commented on zipextimporter.py:
```diff
@@ -42,138 +43,227 @@
+    names_pywin32 = frozenset(["pywintypes", "pythoncom"])
+    verbose = _memimporter.get_verbose_flag()
+
+    def __init__(self, path_or_importer):
```
Could you comment on why you changed the `__init__` API w.r.t. `zipimporter`? I would keep the two API equal, and so stick to use `path` as only argument.

SeaHOH replied:
Needed by monkey patch. Yes, the API was changed, but no problem, argument `zipimporter` be used by `ZipExtensionImporter` itself only.

albertosottile commented on zipextimporter.py:
```diff
@@ -42,138 +43,227 @@
-
-            if sys.version_info >= (3, 0):
-                # name of initfunction
-                initname = "PyInit_" + fullname.split(".")[-1]
-            else:
-                # name of initfunction
-                initname = "init" + fullname.split(".")[-1]
-            filename = fullname.replace(".", "\\")
-            if filename in ("pywintypes", "pythoncom"):
-                filename = filename + "%d%d" % sys.version_info[:2]
-                suffixes = ('.dll',)
+            zipextimporter = self._zipextimporter
+        except AttributeError:
+            self._zipextimporter = zipextimporter = ZipExtensionImporter(self)
+        return zipextimporter
```
All this (plus the `__init__` manipulation) feels like a hack. Could you explain why it was necessary? `zipextimporter` was just a regular child class of `zipimporter`, and I would appreciate to keep it that way... Do you think it would be possible to restore this status? 

In general, I like the new methods provided by this MR, but I do not see why we need to overhaul the class base structure to implement them.

SeaHOH replied:
All this are needed by monkey patch. As I said above, I have made changes, only for improve compatibility. Please imagine that, if someone wrote some code about `zipimporter`, he can run it with `zipextimporter` without any modification.

albertosottile replied:
In the current implementation, if someone wants to extend `zipimporter` they would most likely write a wrapper class around it as we did, inheriting from it. If that someone then wants to extend `zipextimporter`, instead, they would just have to change their parent class.

Conversely, in the implementation proposed in this PR, extending `zipextimporter` would be impossible. 

I appreciate the additions to `zipextimporter` provided in this PR, however I am strongly against abandoning the status of "simple" child of `zipimport`. If you could accommodate your changes without interfering with the child status of `zipextimporter` (no `hasattr` magic, to be more clear) or with the original code of `zipimporter`, that would be great. Otherwise, I am afraid I would have to reject this PR and (if you are not against this) I will  just take the changes that are not disruptive manually from your repository...

SeaHOH replied:
In fact extending `zipextimporter` is possible, but monkey patch is useless in this case, we need install the child to `path_hooks` just like current which the way is always available.

I know the logic complexity has increased, but there has only made addition and no replacement, all old usages were kept. If you agree with this, I will go on to work on issues which you pointed out above.

albertosottile reviewed (changes requested):
I did not finish the whole review, but I already have significant questions. Therefore, I think it is easier if I just submit this first round now.

SeaHOH commented on zipextimporter.py:
```diff
@@ -42,138 +43,227 @@
-
-from importlib import machinery
-from importlib import util
+import _imp
```
I think `_imp` is much different from `imp`, but OK, it will be changed, this is not important.

SeaHOH commented:
I can read English, but does not good at speak and write. If you don’t fully comprehend what I wrote, please tell me, I will try to explain further.

albertosottile commented:
I apologize for the delay, as I did not have much time to work on `py2exe` lately. it looks like we got stuck mainly on two issues:
- monkey patching `zipimport.zipimporter` vs subclassing it
- recoding `zipextimporter` as a standalone class vs making it a regular child of `zipimport.zipimporter`

As I mentioned multiple times, I like the proposed new features and your commitment to this code. But I do not see any benefit into the two points mentioned before. Instead, i only see that it would make much harder to maintain this module, because the code would get much more complex. Furthermore, it would make the code much more brittle, as any monkey patching is bound to break the moment the original code is changed.

In summary, I would really appreciate if you could apply your changes without overhauling the original code structure (this means, leaving `zipextimporter` as a child of `zipimport.zipimporter` and do not monkey patch `zipimport.zipimporter`). If you do not wish to do this, then I am afraid I have to close this PR and, perhaps, in the future I could look at your repository and try to integrate new features from your additions in this code base.

SeaHOH commented:
OK, it will be done in the next few days.

SeaHOH committed:
zipextimporter: improves compatibility

Add support import .pyd packages, e.g. `name\\__init__.pyd.`, also makes
modules search order as same as non-zip.

Add support import pywin32 DLLs.

Add support compatible with export hook name (PEP 489).

Add some missing methods.

Tweak install function.

Minimum compatible to CPython 36.

SeaHOH commented on zipextimporter.py:
```diff
@@ -42,137 +42,196 @@
-            if hasattr(module, '__memimported__'):
+                return super().load_module(fullname)
+
+        spec._set_fileattr = True  # has_location, use for reload
```
Should `spec.has_location` be used as better?

albertosottile replied:
In principle, yes, but we need to make sure this passes the tests.

SeaHOH replied:
Yes, but I forgot why test fail, seems issue is not related this PR. I will push new commit in later time.

SeaHOH committed:
zipextimporter: tweaks

SeaHOH committed:
zipextimporter: tweaks

SeaHOH committed:
zipextimporter: tweaks
